<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Maria Doyle, Bioconductor Community Manager">
<meta name="author" content="Bioconductor Core Developer Team">
<meta name="dcterms.date" content="2023-02-24">

<title>Bioconductor Notes, Winter 2024</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="bioc_files/libs/clipboard/clipboard.min.js"></script>
<script src="bioc_files/libs/quarto-html/quarto.js"></script>
<script src="bioc_files/libs/quarto-html/popper.min.js"></script>
<script src="bioc_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="bioc_files/libs/quarto-html/anchor.min.js"></script>
<link href="bioc_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="bioc_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="bioc_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="bioc_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="bioc_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Bioconductor Notes, Winter 2024</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Maria Doyle, Bioconductor Community Manager </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            University of Limerick
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Bioconductor Core Developer Team </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Dana-Farber Cancer Institute, Roswell Park Comprehensive Cancer Center, City University of New York, Fred Hutchinson Cancer Research Center, Mass General Brigham
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 24, 2023</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Abstract</div>
    <p>We discuss general project news.</p>
  </div>
</div>


</header>


<section id="introduction" class="level1">
<h1>Introduction</h1>
<p><a href="https://bioconductor.org">Bioconductor</a> provides tools for the analysis and comprehension of high-throughput genomic data. The project has entered its twentieth year, with funding for core development and infrastructure maintenance secured through 2025 (NIH NHGRI 2U24HG004059). Additional support is provided by NIH NCI, Chan-Zuckerberg Initiative, National Science Foundation, Microsoft, and Amazon. In this news report, we give some updates on core team and project activities.</p>
</section>
<section id="software" class="level1">
<h1>Software</h1>
<p>The current Bioconductor release is <a href="https://bioconductor.org/news/bioc_3_20_release/">3.20</a>. It is compatible with R 4.4 and consists of 2289 software packages, 431 experiment data packages, 928 up-to-date annotation packages, 30 workflows, and 5 books. are built regularly from source and therefore fully reproducible; an example is the community-developed <a href="https://bioconductor.org/books/release/OSCA/">Orchestrating Single-Cell Analysis with Bioconductor</a>.</p>
<section id="bioconductor-in-python" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-in-python">Bioconductor in Python!</h2>
<p>We’re excited to introduce <a href="https://www.github.com/biocpy">BiocPy</a>, a project aimed at enabling Bioconductor workflows in Python. Analysts today use a variety of languages in their workflows, including R/Bioconductor for statistical analysis and Python for imaging or machine learning tasks. BiocPy aims to facilitate interoperability between R and Python by providing standardized data structures based on existing Bioconductor data structures. These include genomic ranges for interval based operations, summarized experiments and other derivatives for analyzing genomic experiments. To learn more, visit the <a href="https://github.com/biocpy">BiocPy</a> GitHub organization.</p>
</section>
</section>
<section id="core-team-updates" class="level1">
<h1>Core team updates</h1>
<p>What’s new for core?</p>
<section id="partnering-with-outreachy" class="level2">
<h2 class="anchored" data-anchor-id="partnering-with-outreachy">Partnering with Outreachy</h2>
<p>Bioconductor will also participate in Outreachy’s FIXME internship round to offer internships with Bioconductor community projects.</p>
</section>
</section>
<section id="conferences" class="level1">
<h1>Conferences</h1>
<p>Links on conferences?</p>
</section>
<section id="blog" class="level1">
<h1>Blog</h1>
<p>New blog entries to mention?</p>
</section>
<section id="boards-updates" class="level1">
<h1>Boards updates</h1>
<section id="community-advisory-board" class="level2">
<h2 class="anchored" data-anchor-id="community-advisory-board">Community Advisory Board</h2>
<p>Update?</p>
<!--
- The [Community Advisory Board](https://bioconductor.org/about/community-advisory-board/) (CAB) are voting in new members. Outgoing members are Susan Holmes, Leonardo Collado-Torres, Yagoub A. I. Adam, Matt Ritchie, Benilton Carvalho, we gratefully thank them for their service.

- There is a Call for Bioconductor Community-Led events supported by the CAB.
Apply in [this form](https://forms.gle/oTHp26TG3eP9q4eE7)
-->
</section>
</section>
<section id="committees-and-working-groups-updates" class="level1">
<h1>Committees and Working groups updates</h1>
<p>If you are interested in becoming involved with one of these groups please contact the group leader(s).</p>
<p>Updates?</p>
<!--

- [Cloud Methods](https://workinggroups.bioconductor.org/currently-active-working-groups-committees.html#cloud-methods)
  held their first meeting Feb 2023. The working group will meet monthly to
  develop standards and tools for the Bioconductor community related to the
  cloud.

- [Social media working group](https://workinggroups.bioconductor.org/currently-active-working-groups-committees.html#social-media).
  A new group, started in Feb 2023, have set up
  [Bioconductor Mastodon acccount](https://genomic.social/@bioconductor).

- [Teaching Committee](https://workinggroups.bioconductor.org/currently-active-working-groups-committees.html#education)
    - Are planning a hackathon for the
      [Bioconductor Carpentries RNA-seq lesson](https://carpentries-incubator.github.io/bioc-rnaseq/).
    - Will be participating in
      [Galaxy Training Smorgasbord 2023](https://gallantries.github.io/video-library/events/smorgasbord3/),
      a free online global event, May 22-26.

- [Website working group]([https://workinggroups.bioconductor.org/currently-active-working-groups-committees.html#website)
  are planning the redesign of [bioconductor.org](bioconductor.org).

-->
</section>
<section id="other" class="level1">
<h1>Other</h1>
<ul>
<li>Summer school <em>“Statistical Data Analysis for Genome-Scale Biology”</em> <a href="https://csama2023.bioconductor.eu/">CSAMA 2023</a> in Bressanone-Brixen, Italy, June 11-16 2023. Application deadline May 15.</li>
<li>Course <em>“Statistical Analysis of Genome Scale Data”</em> at <a href="https://meetings.cshl.edu/courses.aspx?course=C-DATA&amp;year=23">CSHL</a> in New York, USA, June 30 - July 13 2023. Application deadline March 15.</li>
<li>Workshop <em>“Opportunities for Bioconductor and ELIXIR communities to co-develop training infrastructure”</em> accepted for the <a href="https://elixir-europe.org/events/elixir-all-hands-2023">ELIXIR All Hands meeting</a> in Dublin, Ireland, June 5-8 2023.</li>
<li>Workshop <em>“Orchestrating Large-Scale Single-Cell Analysis with Bioconductor”</em> accepted for <a href="https://www.iscb.org/ismbeccb2023-programme/tutorials#ip4">ISMB/ECCB 2023</a> in Lyon, France, July 23 2023.</li>
</ul>
</section>
<section id="using-bioconductor" class="level1">
<h1>Using Bioconductor</h1>
<p>Start using Bioconductor by installing the most recent version of R and evaluating the commands</p>
<pre><code>  if (!requireNamespace("BiocManager", quietly = TRUE))
      install.packages("BiocManager")
  BiocManager::install()</code></pre>
<p>Install additional packages and dependencies, e.g., <a href="https://bioconductor.org/packages/SingleCellExperiment">SingleCellExperiment</a>, with</p>
<pre><code>  BiocManager::install("SingleCellExperiment")</code></pre>
<p><a href="https://bioconductor.org/help/docker/">Docker</a> images provides a very effective on-ramp for power users to rapidly obtain access to standardized and scalable computing environments. Key resources include:</p>
<ul>
<li><a href="https://bioconductor.org">bioconductor.org</a> to install, learn, use, and develop Bioconductor packages.</li>
<li>A list of <a href="https://bioconductor.org/packages">available software</a> linking to pages describing each package.</li>
<li>A question-and-answer style <a href="https://support.bioconductor.org">user support site</a> and developer-oriented <a href="https://stat.ethz.ch/mailman/listinfo/bioc-devel">mailing list</a>.</li>
<li>A community slack workspace (<a href="https://slack.bioconductor.org">sign up</a>) for extended technical discussion.</li>
<li>The <a href="https://f1000research.com/gateways/bioconductor">F1000Research Bioconductor gateway</a> for peer-reviewed Bioconductor workflows as well as conference contributions.</li>
<li>The <a href="https://www.youtube.com/user/bioconductor">Bioconductor YouTube</a> channel includes recordings of keynote and talks from recent conferences including BioC2022, EuroBioC2022, and BiocAsia2021, in addition to video recordings of training courses.</li>
<li>Our <a href="https://github.com/Bioconductor/Contributions">package submission</a> repository for open technical review of new packages.</li>
</ul>
<p>Upcoming and recently completed events are browsable at our <a href="https://bioconductor.org/help/events/">events page</a>.</p>
<p>The <a href="https://bioconductor.org/about/technical-advisory-board/">Technical</a> and and <a href="https://bioconductor.org/about/community-advisory-board/">Community</a> Advisory Boards provide guidance to ensure that the project addresses leading-edge biological problems with advanced technical approaches, and adopts practices (such as a project-wide <a href="https://bioconductor.org/about/code-of-conduct/">Code of Conduct</a>) that encourages all to participate. We look forward to welcoming you!</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>